spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -528,12 +529,20 @@ public Scan buildChangelogScan() {
           SparkReadOptions.END_TIMESTAMP);
     }
 
+    boolean emptyScan = false;
     if (startTimestamp != null) {
       startSnapshotId = getStartSnapshotId(startTimestamp);
+      if (startSnapshotId == null && endTimestamp == null) {
+        emptyScan = true;
+      }
     }
 
     if (endTimestamp != null) {
-      endSnapshotId = SnapshotUtil.snapshotIdAsOfTime(table, endTimestamp);
+      endSnapshotId = SnapshotUtil.snapshotIdAsOfTime(table, endTimestamp, true);
+      if ((startSnapshotId == null && endSnapshotId == null)
+          || (startSnapshotId != null && startSnapshotId.equals(endSnapshotId))) {
+        emptyScan = true;
+      }
```

Comment:
I'd suggest a new method to generate a snapshot range from the time range when time range is used(there is at least one time range options). This method can return either a valid snapshot range, or empty range. In case of the empty range, we create an empty scan.

Reply:
However, we need to both check snapshot and timestamp to decide whether we create an empty scan. There's also a case where `startSnapshotId.equals(endSnapshotId)`.

Reply:
We should check that even endTimestamp is null, right? Currently, it only checks when endTimestamp != null. With that, we can move the check out of the range conversion. WDYT?